op-superchain/backend.go

```diff
@@ -0,0 +1,163 @@
+			TransactionsCacheSize: 10,
+			HeadersCacheSize:      10,
+			PayloadsCacheSize:     10,
+			RPCProviderKind:       sources.RPCKindAny,
+			MethodResetDuration:   time.Minute,
+		},
+	}
+
+	l2Client, err := sources.NewL1Client(l2Node, log, cacheMetrics, &l2ClientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct l2 client: %w", err)
+	}
+
+	// retrieve the current references before setting up the poll
+	finalizedHeadRef, err := l2Client.L1BlockRefByLabel(ctx, eth.Finalized)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query finalized block ref: %w", err)
+	}
+
+	backend.l2FinalizedBlockRef = &finalizedHeadRef
+	l2FinalizedHeadSignal := func(ctx context.Context, sig eth.L1BlockRef) {
+		backend.mu.Lock()
+		backend.l2FinalizedBlockRef = &sig
+		backend.mu.Unlock()
+	}
+
+	pollInterval, timeout := time.Second*12*32, time.Second*10
+	backend.l2FinalizedHeadSub = eth.PollBlockChanges(log, l2Client, l2FinalizedHeadSignal, eth.Finalized, pollInterval, timeout)
+
+	return &backend, nil
+}
+
+func (b *backend) MessageSafety(ctx context.Context, id MessageIdentifier, payloadBytes hexutil.Bytes) (MessageSafetyLabel, error) {
+	b.log.Info("message safety check", "chain_id", id.ChainId, "block_num", id.BlockNumber, "log_index", id.LogIndex)
+
+	// ChainID Invariant.
+	//   TODO: Assumption here that the configured peers exactly maps to the registered dependency set.
+	//   When the predeploy is added, this needs to be tied to the dependency set registered on-chain
+	l2Node, ok := b.l2PeerNodes[id.ChainId]
+	if !ok {
+		return Invalid, fmt.Errorf("peer with chain id %d is not configured", id.ChainId)
+	}
+
+	var logs []types.Log
+	var header *types.Header
+
+	// Since eth_getLogs doesn't support specifying the log index, we fetch
+	// all the outbox reciepts for this block (TODO: add address filter). The
+	// timestamp is grabbed via the block header as getLogs omits this
+	blockNumber := hexutil.EncodeUint64(id.BlockNumber)
+	filterArgs := map[string]interface{}{"fromBlock": blockNumber, "toBlock": blockNumber}
+	batchElems := make([]rpc.BatchElem, 2)
+	batchElems[0] = rpc.BatchElem{Method: "eth_getBlockByNumber", Args: []interface{}{blockNumber, false}, Result: &header}
+	batchElems[1] = rpc.BatchElem{Method: "eth_getLogs", Args: []interface{}{filterArgs}, Result: &logs}
+	if err := l2Node.BatchCallContext(ctx, batchElems); err != nil {
+		return Invalid, fmt.Errorf("unable to request logs: %w", err)
+	}
+	if batchElems[0].Error != nil || batchElems[1].Error != nil {
+		return Invalid, fmt.Errorf("caught batch rpc failures: getBlockByNumber: %w, getLogs: %w", batchElems[0].Error, batchElems[1].Error)
+	}
+	if header == nil {
+		return Invalid, fmt.Errorf("block %d does not exist", id.BlockNumber)
+	}
+
+	// Message Log Integrity
+	// 	 -- BlockNumber & ChainID are handled via the RPC connection & inputs
+
+	// TODO: If we filter by address, then this needs to change
+	if id.LogIndex >= uint64(len(logs)) {
```

Comment:
We should be sure to break this out into a library so it can be tested and reused independently

Reply:
yea agreed, this is main reason why I didn't have this coupled the rpc server so that it can constructed and held reference to as lib. How the subscriptions and config parameters are created will be better understood when integrating this into the verifier codebase in op-node